### DISCUSSION TITLE
Accessibility Request

### DISCUSSION THREAD

**Discussion by reaperaccessible** [jamulussoftware/jamulus#3386] [General]
Hi,

First I would like to congratulate and thank all the developers for Jamulus. It really is the best solution for jamming online.

This is to ask you for some small improvements regarding accessibility with screen readers software like NVDA.
I am a blind user using NVDA latest version, the screen review software with the latest version of Windows 10 pro.
Jamulus is completely accessible and I thank you for that.
But there are 2 points that I would like to explain to you.

1. It is possible to know our ping but for some reason we do not have access to the delay and jitter.
These 2 pieces of information are very important and blind users do not have access.

2. When I press Ctrl+H to open chat, Tab and Shift+Tab don't work properly with NVDA, I have to press several times on Tab or Shift+Tab before NVDA says which part I'm in.

3. When a user writes in the chat, the message is automatically read by NVDA. This is disturbing when I'm jam. Would it be possible to add a checkbox to completely disable chat? If I could disable the chat so that the messages no longer appear, that would solve this problem.

Do not hesitate to contact me if you would like me to test a beta version.

Thanks to consider this request.

**Comment by ann0see** [comment]
Hi @reaperaccessible,

Thanks for opening this discussion. @chigkim knows more about accessibility. 

> It is possible to know our ping but for some reason we do not have access to the delay and jitter.
> These 2 pieces of information are very important and blind users do not have access.

This sounds like a bug. Definitely this shouldn't happen. Could you please open an issue in the issue tracker for this? I don't see anything obvious from the code unfortunately.

> When I press Ctrl+H to open chat, Tab and Shift+Tab don't work properly with NVDA, I have to press several times on Tab or Shift+Tab before NVDA says which part I'm in.

Probably it jumps over some unlabelled (?) buttons (Maybe the text input box and then send). This needs some closer investigation for sure.

> When a user writes in the chat, the message is automatically read by NVDA. This is disturbing when I'm jam. Would it be possible to add a checkbox to completely disable chat? If I could disable the chat so that the messages no longer appear, that would solve this problem.

It's also sometimes annoying for people with vision. The Chat dialog may just pop up right in front of your session. This is something which needs more thought and not a quick fix.


Related: https://github.com/jamulussoftware/jamulus/issues/3315

**Reply by ann0see** [reply]
Update: I have now tested NVDA on Windows 11 with Jamulus 3.11.0. If you hover over the Delay label the screen reader reads the delay in MS to me. If you hover over the LED it correctly says if it's red or green. I cannot reproduce this issue unfortunately. I cannot use tab (or don't know how to use it to tab through all those stats. Maybe this doesn't work correctly).

**Reply by reaperaccessible** [reply]
Hi,
NVDA may show you the information correctly if you hover over the information. The problem is that we can't use the mouse.
However, the good news is that if NVDA can indicate what the mouse is hovering over, there is not much missing to make this possible with the keyboard.
Ideally, all information should be available with the up, down, left, right arrows, Tab and Shift+Tab.
Another solution, a shortcut which tells us the value.
If I select a server from the list, would it be possible to send the information to the screen reader via a keyboard shortcut?
Thank you for your time on this issue.

**Reply by ann0see** [reply]
Can you tell me how you get to the ping value via keyboard? 
I cannot select much by using tab only. 
The docuentation says that insert is the NVDA button, but I still can't navigate the UI by pressing that and the arrow buttons.

**Reply by reaperaccessible** [reply]
You must first press the button above 7 on the numeric keypad in order to be in num lock disabled.
You select a server from the list and press zero on the keypad and hold and use 4 and 6 on the keypad to navigate by object.
This way of navigating is useful when an application does not exactly meet the accessibility standard. This allows us to get out of the box in order to navigate on our own in order to discover certain options that appear on the screen which are not easily accessible.
So if you select a server and want to know the ping, the shortcut to use is NVDA+6 from the numeric keypad. The NVDA key is pad zero.
Don't forget to put yourself in Num Lock Off mode by pressing the button above the 7 on the number pad.

**Reply by ann0see** [reply]
I still don't succeed using the screen reader unfortunately. If I press the respective buttons on the number pad, it does go through the mixer board (at least the title and the faders?), but I can't come out of it. However - since it's a user issue on my side, I've shot into the blue with 

https://github.com/ann0see/jamulus/pull/160

Download the Windows artifact from 
https://github.com/ann0see/jamulus/actions/runs/11150598783 as soon as it's finished. I have added tabstops for ping/delay etc. I don't think that's the final solution, but maybe we can build from this.

**Reply by reaperaccessible** [reply]
Ok, I will give you all the steps to follow the procedure exactly with NVDA with Keyboard.
1. Start NVDA.
2. Open Jamulus with the keyboard or mouse.
3. Press Ctrl+C to open the server connection window.
4. Press Tab once to focus the Directory combo box, drop-down list to select Any Genre 1 or 2.
5. Use the up or down arrow keys only to select Any gender 1 and 2.
6. Once Any gender 1 or 2 is selected, if you press Tab once, you will now be in the edit field for the filter.
7. If you press Tab again, you are now sure Show all musicians check box.
8. If you press Tab again, you are now in the server drop-down list, Server list view 
9. Now up arrow and down arrow allows you to select a server from the list.
By default, you will find yourself at the start of the list.
So, you can simply press Down Arrow once then Up Arrow to select the first server in the list.
You will hear the name of the server spoken by NVDA that you see selected on the screen.
If you press Enter, you confirm your choice and enter this server.
But for now let's stay in the list of servers with the first one available in this list and continue.
10. Once the server is selected in the list, you press the button at the top of 7 on the number pad to put the keypad in Num Lock Off mode. It's a toggle button so make sure you're in Num Lock Off mode otherwise the following won't work.
11. Press and hold the zero button on the numeric keypad. When you hold this button, you are not supposed to hear the Windows bell. If this is the case, the numeric keypad is not in the correct mode.
12. While the zero button is pressed, press 6 on the pad once to move the focus of NVDA by object to the right.
You should hear NVDA stated something like this:
3 ms Ping Time column 2 level 1
13. If you hold down the zero key on the keypad and press 6 on the keypad again, you should now hear something like this:
0/8 Musicians column 3 level 1
14. If you hold down zéro and press 6 again, you should hear this:
Central, Canada Location column 4 level 1
As you can see, NVDA manages to read the first 4 columns and that's it. I don't know if there are more than 4 but if so, that's where the problem comes from.
This procedure should work on your machine since I do this every day on Jamulus.
But if you have other questions or need to clarify certain points, don't hesitate.
Thanks my friend.

**Reply by reaperaccessible** [reply]
Update:
A musician friend on Jamulus told me an important detail that I didn't know.
I thought it was possible to see the delay in the connection window while everything happens in the main interface when the connection to a server is established.
So, I downloaded your dev version and installed it. Unfortunately I can't test it because as soon as I focus the list of available servers, Jamulus freezes and I have to force it to quit.
I uninstalled it and reverted to the stable version and everything came back as before.

**Reply by ann0see** [reply]
Ok. Thanks. Now I could follow your instructions. Indeed there are only the four columns in the connection dialog. We cannot calculate the overall delay if there is no connection. 

Strange that my version starts hanging. On my machine everything is fine. Maybe it has to do something with your audio device?

Would you say that 1. is therefore ok? Can you get to the Delay/Jitter in the main mixer board?

2. If I tap through the chat dialog, I cannot reproduce your issue. However, if I connect to low latency band, the screen reader does not say anything if I tap over links in the welcome message. This may be the issue?
3. This is also annoying for users with vision - so the issue is broader.

**Reply by reaperaccessible** [reply]
Ok, I managed to install it and everything works correctly.
On the other hand, I don't see any change in the main window when I browse the elements with Tab.
I'll tell you exactly what I'm doing and you'll tell me if what I can see with NVDA is normal or not.
If I select any server in the connection window and press Enter, I access the server.
By default, it is Connect and disconnect toggle button which is selected.
So if we go from there and I press Tab, the next item is Mute Myself, next is Settings, Chat, Reverb effect level setting, Personal Mix at: reaperaccessible, Local mix level setting of the current audio channel at the server, 50  50  <b>Panning:, Mute Lee (my self), Solo Lee Drums, Group Lee Drums & Keys, Drum Set, and I come back toConnect and disconnect toggle button.
So, I don't see any Ping, Delay or local jitter status information with Tab.
Am I doing something wrong?
Thanks again.

**Reply by ann0see** [reply]
This is the same as before. I wanted to include a way to be able to tab to the LEDs or values for Delay, Jitter and Ping shown in the main mixer board. 
However, it seems as if you cannot just add a tabstop to those labels as they cannot be clicked. 

How do you access the ping/delay/jitter LEDs on the mixer? Is that even possible?

**Reply by chrisrimple** [reply]
It's been a very long time since I last worked with a screen reader.  But back then, it was possible to advance through each element in a dialog, including labels (and not just controls), and have them read.  Perhaps that's still possible, even without tab stops?

**Reply by reaperaccessible** [reply]
It is actually not possible to focus on a label. The most common practice is to use a read-only edit field, previously populated with data.

**Comment by chigkim** [comment]
Sorry everyone, I haven't really kept up with Jamulus after the pandemic.
Having said that, the bug for the server list is from QT TreeWidgetItem.
I had a very very hacky workaround where it cast every label as button, so screen readers can focus and read it, but that modifies the UI for everyone not just screen reader users. That was one of the main reasons why I had my own forked repo for a while until QT partially fixed their bug so screen readers can at least read the name of servers.
I can't promise anything since my table is pretty full at the moment, but I'll try to look over the weekend and see anything changed from QT, and if there's anything we can do about it.

**Reply by reaperaccessible** [reply]
Cool, thanks a lot for your help guys.
